File: android/app/src/main/java/kz/qpvpn/vpn/LinkWatch.kt
package kz.qpvpn.vpn

/**
 * Сторож связи: замечает, что туннель поднят, а сервера за ним уже нет.
 *
 * Признак один: телефон шлёт, а в ответ тишина. Именно так выглядит переезд
 * с Wi-Fi на мобильную сеть — туннель на месте, пакеты уходят, обратно не
 * приходит ничего.
 *
 * По времени последнего рукопожатия это определять нельзя. Лежащий в кармане
 * телефон не обновляет сессию, потому что ему нечего слать, — и приложение
 * поднимало туннель заново каждые две минуты простоя. Связь при этом каждый
 * раз рвалась на секунду-другую, и всё казалось тормозным на ровном месте.
 */
class LinkWatch(
    /** Столько ждём ответа, пока шлём, прежде чем считать связь оборванной. */
    private val silenceMillis: Long = 45_000,

    /** Не чаще этого пересоздаём туннель, чтобы не биться в закрытую дверь. */
    private val repairPauseMillis: Long = 120_000,
) {

    private var lastRx = 0L
    private var lastTx = 0L
    private var rxMovedAt = 0L
    private var lastRepair = 0L

    /** Туннель только что поднялся: считаем отсчёт заново. */
    fun start(rx: Long, tx: Long, now: Long) {
        lastRx = rx
        lastTx = tx
        rxMovedAt = now
        // Пауза между починками отсчитывается от прошлой починки, а её не
        // было: иначе первую пришлось бы ждать лишние две минуты.
        lastRepair = now - repairPauseMillis
    }

    /**
     * Пора ли поднимать туннель заново.
     *
     * Возвращает true не чаще, чем раз в [repairPauseMillis]: если сервер
     * недоступен совсем, бесконечные попытки только сажают батарею.
     */
    fun stalled(rx: Long, tx: Long, now: Long): Boolean {
        if (rx > lastRx) rxMovedAt = now
        val sending = tx > lastTx
        lastRx = rx
        lastTx = tx

        if (!sending) return false
        if (now - rxMovedAt < silenceMillis) return false
        if (now - lastRepair < repairPauseMillis) return false

        lastRepair = now
        return true
    }
}

File: Sources/KupibasCore/LinkWatch.swift
import Foundation

/// Сторож связи: замечает, что туннель поднят, а сервера за ним уже нет.
///
/// Признак один: компьютер шлёт, а в ответ тишина. Именно так выглядит
/// переезд с Wi-Fi на другую сеть или пробуждение из сна — интерфейс на
/// месте, пакеты уходят, обратно не приходит ничего.
///
/// По времени последнего handshake это определять нельзя. Простаивающий
/// туннель сессию не обновляет, потому что ему нечего слать, — и служба
/// пересоздавала его каждые три минуты простоя. Связь при этом рвалась,
/// открытые соединения обрывались, и всё казалось тормозным на ровном месте.
public final class LinkWatch {

    /// Столько ждём ответа, пока шлём, прежде чем считать связь оборванной.
    private let silence: TimeInterval

    /// Не чаще этого пересоздаём туннель, чтобы не биться в закрытую дверь.
    private let repairPause: TimeInterval

    private var lastRx = 0
    private var lastTx = 0
    private var rxMovedAt: TimeInterval = 0
    private var lastRepair: TimeInterval = 0

    public init(silence: TimeInterval = 45, repairPause: TimeInterval = 120) {
        self.silence = silence
        self.repairPause = repairPause
    }

    /// Туннель только что поднялся: считаем отсчёт заново.
    public func start(rx: Int, tx: Int, now: TimeInterval) {
        lastRx = rx
        lastTx = tx
        rxMovedAt = now
        // Пауза между починками отсчитывается от прошлой починки, а её не
        // было: иначе первую пришлось бы ждать лишние две минуты.
        lastRepair = now - repairPause
    }

    /// Пора ли поднимать туннель заново.
    public func stalled(rx: Int, tx: Int, now: TimeInterval) -> Bool {
        if rx > lastRx { rxMovedAt = now }
        let sending = tx > lastTx
        lastRx = rx
        lastTx = tx

        guard sending else { return false }
        guard now - rxMovedAt >= silence else { return false }
        guard now - lastRepair >= repairPause else { return false }

        lastRepair = now
        return true
    }
}
